### DISCUSSION TITLE
Tagged Union field validation results in the wrong error messages.

### DISCUSSION THREAD

**Discussion by ojii** [fastapi/fastapi#6543] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
I've written a test which shows the problem:

https://github.com/ojii/fastapi/blob/tagged-union-validation/tests/test_tagged_union_field_validation_errors.py
```

### Description

Given ```python
class A(BaseModel):
    tag: Literal["a"]

class B(BaseModel):
    tag: Literal["b"]
    value: conint(ge=0, le=1)

class Item(BaseModel):
    value: Union[A, B]
```

If you send `{value: {tag: 'b', value: 2}}` the error message is:

```python
{
    "detail": [
        {
            "loc": ["body", "value", "tag"],
            "msg": "unexpected value; permitted: 'a'",
            "type": "value_error.const",
            "ctx": {"given": "b", "permitted": ["a"]},
        },
        {
            "loc": ["body", "value", "value"],
            "msg": "ensure this value is less than or equal to 1",
            "type": "value_error.number.not_le",
            "ctx": {"limit_value": 1},
        },
    ]
}
```

I think it should be:

```python
{
    "detail": [
        {
            "loc": ["body", "value", "value"],
            "msg": "ensure this value is less than or equal to 1",
            "type": "value_error.number.not_le",
            "ctx": {"limit_value": 1},
        },
    ]
}
````

### Operating System

macOS

### Operating System Details

(os should not matter)

### FastAPI Version

0.68.0 (25a4828)

### Python Version

3.8.3 and 3.9.1

**Comment by ojii** [comment]
sorry, this is a bug in pydantic. will report it there.

**Comment by YuriiMotov** [comment, marked as answer]
This is not a bug, but just how it works.
Pydantic tries to validate data using all types in Union and collect all error messages.

You can make error message cleaner using discriminated unions: https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions
